### PR TITLE
Adding a guide to setup CONSUL on Heroku

### PR DESCRIPTION
**Pull request by PierreMesure**
Hey,

I wrote a quick tutorial on how to deploy Consul on Heroku. It doesn't include features like the mail server that I haven't had the time to explore yet.

For now, I'm just submitting a pull request to get some feedback on the process. If this manual version is OK with you, I'd like to create a branch that includes all the file changes to make the process even shorter.

### CODE REVIEW

**Comment by bertocq** [comment]
Thanks @PierreMesure !! I've read it and seems awesome :) (no time to try it but I will eventually).

I'm only concerned about the recommendation to commit `config/secrets.yml` to the version control system :( Since for sure there will be sensitive data on any consul installation. But I know right now there is no other alternative.. Maybe we should support https://github.com/bkeepers/dotenv or just recommend using it at the `Gemfile_custom` file 🤔. Many fork maintainers are not used to work on public repositories (we've had at AyuntamientoMadrid a repo with plain text users/passwords/hosts for months :/ because no-one realized that the repo was going to be "opensourced"). Do you think you can give it a try to see if using the gem works without too much hassle?

About the branch you mention, I'm not sure I understand what you mean.. a branch from consul/consul with:
- new heroku.yml file with default recommended config
- changes at Gemfile_custom & Gemfile.lock
- config/secrets.yml file 

So anyone can just download consul/consul and switch from #master branch to #heroku branch and go on from there... right?

If that's so...  I think we could just integrate in master:
- a new `heroku.yml.example` file (with a `cp heroku.yml.exampe heroku.yml` tip on the guides where you're recommending creating it) 
- the changes at `Gemfile_custom` but commented out and with a previous comment `# Uncomment following lines if you're deploying to heroku` & a clear reference to that action on the heroku guidelines. 
- But the secrets.yml change at #master branch would be a no go as many forks wouldn't understand that they need to remove it and avoid commiting it if they're not using heroku.

**Comment by PierreMesure** [comment]
Thanks for the quick answer!

Feedback on the secrets.yml was exactly what I wanted!

Funny anecdote, I got my [first consul](https://github.com/DinRiksdag) working without committing that file! I just added a line in production.rb and it works fine! But when writing the tutorial, I couldn't reproduce it and I read somewhere in the Heroku doc that they expected the secrets.yml to be committed and filled with ENV variables.

I can give dotenv a try, [figaro](https://github.com/laserlemon/figaro) also seems like a popular alternative. [heroku_secrets](https://stackoverflow.com/questions/21544626/how-do-you-manage-secret-keys-and-heroku-with-ruby-on-rails-4-1-0beta1/22458102#22458102) seems like an even simpler solution if you want to avoid moving your values around. If you were to include one of these three in your releases in the future, which one would you choose? dotenv seems to have the best support.

About the branch, it's exactly what I had in mind (without the secrets.yml if the fix mentioned above works) as a first step.
Second step would be to add a [Heroku button](how to create heroku button) to make it even simpler. My question is how much would be automated in that button. I have very little knowledge in app deployment and with Consul but I've seen people include addons such as memcached, Redis, etc. Maybe we can start from a simple app with a database and a mail server and build from there?

**Comment by PierreMesure** [comment]
I tried figaro first because the dotenv doc explicitly says that they are not meant to be used to deploy to production. It works great but it doesn't remove the requirement that heroku has to see your secrets.yml in the repo.

What I did is I copied and pasted the secrets.yml file into the newly created application.yml. I then replaced all the real variables in secrets.yml with shortcuts to ENV variables with the same name and I used `bundle exec figaro heroku:set -e production` to set all these variables as configuration variables in production.

I think it would be great if you could integrate this in your release as it would make it much easier for people that will deploy on platforms such as Heroku. Moreover, it would require only little adaptation from people used to putting their secrets in secrets.yml as application.yml can keep the same layout. But I understand it's not an insignificant choice.

**Comment by PierreMesure** [comment]
I managed to get heroku_secrets working as well, and with secrets.yml uncommitted. You decide what you want to adopt when you have time to look into it. The problem with heroku_secrets seems to be that it's unsupported and the rake task to send the env variables into production doesn't seem to work anymore...

**Comment by PierreMesure** [comment]
I just found out that adding the line

`config.secret_key_base = ENV['SECRET_KEY_BASE']`

at the top of _config/environment/production.rb_ works like a charm. The trick is that it doesn't work when you write it at the bottom of the file, which is why I had it working on one app but not on the other.

Maybe you could include that line in a release with a condition? Something like if no secrets.yml check that ENV variable? This would avoid disrupting the installation of people having the file in production while making it work on Heroku.

**Comment by bertocq** [comment]
Hey @PierreMesure sorry for the delay to answer!

I've read about figaro and seems as unmaintained as heroku_secrets, but also doesn't offer a solution to the problem we've.

I would describe the underlaying problem to be solved as: **"We need consul to read secret values from either ENV or secrets.yml when calling Rails.application.secrets.secret_variable_name"** so when app is ran on heroku it uses ENV but when its any other server it can keep using secrets.yml (without being commited to git)

# Solution A: Secrets class
The first solution that came to mind was to create a `lib/secrets.rb` class that "encapsulates" the access to secrets. That would mean replacing all existing `Rails.application.secrets.secret_variable_name` calls for `Secrets.secret_variable_name`. And the Secrets class would try to find if there is a `ENV['SECRET_VARIABLE_NAME']` set, or try to fallback to `Rails.application.secrets.secret_variable_name`. But this seems somehow odd or more like a bad practice... since any dev used to Rails conventions wouldn't know that on Consul you can't just use `Rails.application.secrets*` because we've this fallback behaviour on Secrets class 🤔 

# Solution B: Altering Rails.application.secrets
So as we where saying on solution A, we don't want to break existing conventions... what about changing the behaviour of Rails.application.secrets ? Well maybe that's even worse as upgrading rails version would make us run into conflicts or some scenarios where the fallback logic could get in trouble.

# Solution C:
Change `config/secrets.yml` so instead of values it holds references to ENV variables, but avoid committing it to the git repo.

Like from:
```yaml
default: &default
  secret_key_base: 56792feef405a59b18ea7db57b4777e855103882b926413d4afdfb8c0ea8aa86ea6649da4e729c5f5ae324c0ab9338f789174cf48c544173bc18fdc3b14262e4
```
to
```yaml
default: &default
  secret_key_base: <%= ENV["SECRET_KEY_BASE"] %>
```

Existing installs could just ignore this changes and keep using the `config/secrets.yml` file in their servers untouched... but for new installs it could be a bit strange to ask them to set ENV variables as well as upload the secrets file. Also Heroku installs still would need to commit secrets file to git repo (although it wouldn't contain any values)


# Solution D: special secrets.yml & .env files

We would create two new files: `.env.heroku.example` & `config/secrets.yml.heroku.example`. The `config/secrets.yml.heroku.example` file would have content like `secret_variable_name: <%= ENV['SECRET_VARIABLE_NAME'] %>` (so it would be different from `config/secrets.yml.example`) 

## For Non-Heroku installs:
It would be same as usual: rename existing `config/secrets.yml.example` to `config/secrets.yml` and so on...

## Heroku install steps
1. Rename `*.heroku.example` files:
```bash
cp .env.heroku.example .env
cp config/secrets.yml.heroku.example config/secrets.yml
```
2. Remove `config/secrets.yml` from .gitignore file:
```bash
sed -i '' '\/config\/secrets.yml/d' ./.gitignore
```
3. Commit the `config/secrets.yml` file to git
```bash
git add config/secrets.yml
git commit -m "Add secrets.yml with ENV refs instead of values"
```
4. Add `dot_env` gem to Gemfile_custom to be able to load `.env` values on development (localhost ) and tests environments... so `config/secrets.yml` file can read them.
```
echo "group :development, :test do" >> Gemfile_custom
echo "  gem 'dotenv-rails', '~> 2.2.1'" >> Gemfile_custom
echo "end" >> Gemfile_custom
```

5. Set variables on heroku servers via command line or interface:
https://devcenter.heroku.com/articles/config-vars

6. Set variables on Travis or the CI service of choice
https://docs.travis-ci.com/user/environment-variables/

# Conclusion
From my point of view, last solution is best solution because:
- Doesn't involve any code changes at all
- Doesn't affect in any way to "usual" consul installs (new or existent)
- Allows heroku installs to work without risking exposure of secret values

And drawbacks are:
- Having to make very clear install guides, explaining what those `.env.heroku.example` & `config/secrets.yml.heroku.example` files are for
- Having to maintain parity between `config/secrets.yml.example` variables & values and values at `.env.heroku.example` & variables at`config/secrets.yml.heroku.example`. Although we rarely add one variable each 1-2 releases.
- It's just theory as I haven't had time to test it myself 🙈

**Comment by voodoorai2000** [comment]
Hey @PierreMesure 
Thanks for helping us out here 😃 
Nice explanations @bertocq 😌 
+1 to using special secrets.yml and .env files 👌

**Comment by bertocq** [comment]
Just to make the suggested Solution D easier to understand, this are the two heroku example files I'm referring to with `.env.heroku.example & config/secrets.yml.heroku.example`

https://github.com/consul/consul/commit/e62941458967d68bada93bab87f62671ff67ecce

So just if you checkout `heroku_install` branch you can try it out

**Comment by PierreMesure** [comment]
Hey,

Sorry, I'm only seeing this now! I think it's a great solution! When are you planning to do it?

In addition to this, there are 3 gems that are recommended by Heroku:
- puma (not sure why you're not including it by default as it seems to be the best)
- rails_12factor
- heroku-deflater

As the last two are not necessary for a non-heroku install, I guess it's still worth it to keep them in the separate heroku-install branch.

**Comment by PierreMesure** [comment]
Just used the tutorial again to deploy, it works great except for two points:
- part 5 is wrong. As the heroku container doesn't work with consul (for some reason, assets precompiling cannot be done inside of it because it cannot access a database). These infos are to be put in a Procfile, not in a heroku.yml
- part 8 is also wrong, you can migrate with Heroku CLI, but for some reason you can't write the dev seeds. A beginning of answer lies [here](https://devcenter.heroku.com/articles/rake). I use a workaround by putting the Heroku Postgres database URL as the default in the secrets file and running the rake command from my computer.

Will update the tutorial next week.

File: en/getting_started/deploying-on-heroku.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# Deploying on Heroku
+
+## Manual deployment
+
+This tutorial assumes that you have already managed to clone Consul on your machine and gotten it to work.
```
Comment (aitbw) [inline en/getting_started/deploying-on-heroku.md:5]: Replace `Consul` with `CONSUL`

File: en/getting_started/deploying-on-heroku.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+1. First, create a Heroku account if it isn't already done.
+2. Install the [Heroku CLI](https://devcenter.heroku.com/articles/heroku-cli) and sign in using
+
+   ```
+   heroku login
+   ```
+
+3. Go to your consul repository and instantiate the process
```
Comment (aitbw) [inline en/getting_started/deploying-on-heroku.md:14]: Replace `consul` with `CONSUL`

File: en/getting_started/deploying-on-heroku.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+   If _your-app-name_ is not already taken, Heroku should now create your app.
+
+4. Create a database using
+
+   ```
+   heroku addons:create heroku-postgresql
+   ```
+
+   You should now have access to an empty Postgres database whose address was automatically saved as an environment variable named _DATABASE\_URL_. Consul will automatically connect to it when deployed.
```
Comment (aitbw) [inline en/getting_started/deploying-on-heroku.md:31]: Replace `Consul` with `CONSUL`

**Review by aitbw — APPROVED** [review]
Hey @PierreMesure, thanks a lot for this PR! I'm planning to deploy a CONSUL instance to Heroku to toy around and this guide is amazing.

I've left you a few comments, just to keep everything as consistent as possible :smile_cat:

**Comment by PierreMesure** [comment]
Thanks @aitbw, your comments are actually a good occasion to finish and merge that tutorial. Since you're attacking Heroku, would you be interested in helping?

A big part that I never got the time to write is the mail server and that's precisely what I'm working on right now. Unfortunately, @voodoorai2000's support was erased in Slack.
I've been using two dynos for some time but I'm trying [workless](https://github.com/lostboy/workless) to save one. That plus this doc on [delayed_job](https://devcenter.heroku.com/articles/delayed-job) should be a good starting point.

**Comment by PierreMesure** [comment]
It's interesting to see that the discussion we had about pushing `secrets.yml`or not seems irrelevant now. I've got at least 50 env variables on each of my Heroku app and it seems to be the norm... 😁

File: en/getting_started/deploying-on-heroku.md (hunk 2 above)
Comment (PierreMesure) [reply en/getting_started/deploying-on-heroku.md:14]: Debatable, the main repository (and its forks too, usually) is named 'consul/consul' 😛

Comment (aitbw) [reply en/getting_started/deploying-on-heroku.md:14]: :thinking: fair enough, I think we haven't quite decided which should be the norm for this :stuck_out_tongue:

That being the case, you could disregard my comments altogether :bowing_man:

File: en/getting_started/deploying-on-heroku.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+5. Add a file name _heroku.yml_ at the root of your project and paste the following in it
+
+   ```
+   build:
+     languages:
+       - ruby
+     packages:
+       - imagemagick
+   run:
+     web: bundle exec rails server -e ${RAIL_ENV:-production}
```
Comment (aitbw) [inline en/getting_started/deploying-on-heroku.md:42]: Is this a typo? Shouldn't it be `RAILS_ENV`?

**Comment by voodoorai2000** [comment]
This is very useful @PierreMesure!

We want to setup comprehensive instructions for Heroku and start using it ourselves as a staging environment. Thinking of using the feature of deploying an app per branch too.

Let us know if you keep working on this, we'll do the same 😌

**Comment by PierreMesure** [comment]
@voodoorai2000 I'd love to continue working on it. These past two weeks have been a bit rough but the current one might leave me more time. I'll try to update what I did. In any case, contributions are welcome.

**Comment by voodoorai2000** [comment]
Awesome! Thank you so much @PierreMesure!

**Comment by decabeza** [comment]
Thank you @PierreMesure! 🙏 😌

**Comment by PierreMesure** [comment]
Are we merging this now? It's not finished! Anyway, I think it's better! I'll finish it in January when I have time and I'll add a version for Digital Ocean now that I know how to deploy with Ansible and Capistrano!

**Comment by decabeza** [comment]
Don't worry @PierreMesure we can complete and improve documentation in next PRs! 😉

**Pull request merged**